### PR TITLE
Tracky Refactoring

### PR DESCRIPTION
**Pull request by CuddlyBunion341**
TICKET-19656

The goal is to improve tracky by refactoring existing code and updating the UI.
The main concern on the Plugin is the session management and the ocassionally useless abstractions.

Doc: https://docs.google.com/document/d/1BCTawEQqfiAzCoRx3yih0WRGDkOl_HR8Em61sTzzOds/edit#heading=h.nkew4uhlsarb

<details>
<summary>Updated Layout (without hints)
</summary>

CleanShot 2024-05-24 at 12 20 52@2x

</details>

### CODE REVIEW

File: app/controllers/time_tracker_controller.rb [hunk 1]
```diff
@@ -1,50 +1,48 @@
...
-# rubocop:disable Style/IfUnlessModifier, Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity, Metrics/AbcSize, Metrics/MethodLength
+# rubocop:disable Metrics/AbcSize
+# rubocop:disable Metrics/MethodLength
 class TimeTrackerController < TrackyController
-  def start
-    # Early return if timer already running
-    if @current_timer_session
-      render :update, layout: false and return
-    end
+  def upsert
```
Comment (CuddlyBunion341) [inline app/controllers/time_tracker_controller.rb:6]:
```suggestion
  def create_or_update
```

File: app/controllers/time_tracker_controller.rb [hunk 2]
```diff
@@ -1,50 +1,48 @@
...
+    @current_timer_session = TimerSession.active.find_by(user: User.current)
 
-    @current_timer_session = SessionCreator.new(@current_user, timer_params, params[:commit]).create
+    return start_timer unless @current_timer_session
+    return cancel_timer if params[:cancel]
```
Comment (CuddlyBunion341) [inline app/controllers/time_tracker_controller.rb:10]:
```suggestion
    return cancel_timer if params[:cancel].present?
```

Comment (CuddlyBunion341) [reply app/controllers/time_tracker_controller.rb:10]: Keep cancel as separate endpoint? Look into TimerSession attributes for cancelling.

Follow Rails conventions for params.

File: app/controllers/time_tracker_controller.rb [hunk 3]
```diff
@@ -1,50 +1,48 @@
...
+
+    stop_timer if timer_params[:timer_end].present?
+  end
+
+  private
+
+  def start_timer
+    @current_timer_session = SessionCreator.new(User.current, timer_params, params[:commit]).create
```
Comment (CuddlyBunion341) [inline app/controllers/time_tracker_controller.rb:18]: Refactor params[:commit]??

Comment (CuddlyBunion341) [reply app/controllers/time_tracker_controller.rb:18]: Command params hash

Comment (CuddlyBunion341) [reply app/controllers/time_tracker_controller.rb:18]: Investigate ActiveModel::Attributes

File: app/controllers/time_tracker_controller.rb [hunk 4]
```diff
@@ -1,50 +1,48 @@
...
       render :update, layout: false and return
     end
 
-    # Early close the timer session if the timer_end is set and it's valid
-    if @current_timer_session.session_finished? && @current_timer_session.update(finished: true)
-      TimeSplitter.new(@current_timer_session, @current_timer_session.issues).create_time_entries
-      flash[:notice] = l(:notice_successful_update)
-      render :stop, layout: false and return
+    if @current_timer_session.timer_end.present?
+      stop_timer
```
Comment (CuddlyBunion341) [inline app/controllers/time_tracker_controller.rb:33]: Idea: All render calls happen inside action?

File: app/controllers/time_tracker_controller.rb [hunk 5]
```diff
@@ -1,50 +1,48 @@
...
-    if @current_timer_session.nil?
-      render :stop, layout: false and return
-    end
-
-    if params[:cancel]
-      @current_timer_session.destroy
-      render :cancel, layout: false and return
-    end
+  def cancel_timer
+    @current_timer_session.destroy
```
Comment (CuddlyBunion341) [inline app/controllers/time_tracker_controller.rb:41]: Missing errorhandling

File: app/controllers/time_tracker_controller.rb [hunk 6]
```diff
@@ -1,50 +1,48 @@
...
+    render :cancel, layout: false and return
```
Comment (CuddlyBunion341) [inline app/controllers/time_tracker_controller.rb:42]:
```suggestion
    render :cancel, layout: false
```

Comment (CuddlyBunion341) [reply app/controllers/time_tracker_controller.rb:42]: Add respond to blocks.

File: app/models/timer_session.rb [hunk 7]
```diff
@@ -47,35 +47,35 @@ def relevant_issues
...
-  def limit_recorded_hours
+  def validate_limit_recorded_hours
     return if timer_start.blank? || timer_end.blank?
 
     validate_day_limit
     validate_session_limit
   end
 
-  def enough_time
+  def validate_enough_time
```
Comment (CuddlyBunion341) [inline app/models/timer_session.rb:78]:
```suggestion
  def validate_minimal_duration
```

File: app/views/shared/_js_variables.html.erb [hunk 8]
```diff
@@ -2,7 +2,7 @@
   var RedmineTracky = RedmineTracky || {};
   RedmineTracky.datetimeFormat = "<%= j(I18n.t('timer_sessions.formats.jquery_datetime_format')) %>";
   RedmineTracky.issueCompletionPath = "<%= j(completion_issues_path) %>";
-  RedmineTracky.trackerUpdatePath = "<%= j(update_time_tracker_path) %>";
+  RedmineTracky.trackerUpdatePath = "<%= j(time_tracker_upsert_url) %>";
```
Comment (CuddlyBunion341) [inline app/views/shared/_js_variables.html.erb:5]: Update path name

File: assets.src/src/redmine-tracky/controllers/timer-controller.ts [hunk 9]
```diff
@@ -52,17 +52,10 @@ export default class extends Controller {
...
-      sign +
-      ['hours', 'minutes', 'seconds']
-        .map((v) => timeDiff[v as keyof TimeDiff])
-        .map((v) => Math.abs(Math.floor(v)))
-        .filter((v, i) => i !== 0 || v !== 0) // Remove hours if zero
-        .map((v) => v.toString().padStart(2, '0'))
-        .join(':')
-    )
+    const duration = Duration.fromObject(timeDiff)
+    const formattedDuration = duration.toFormat('hh:mm:ss').replaceAll('-', '')
```
Comment (CuddlyBunion341) [inline assets.src/src/redmine-tracky/controllers/timer-controller.ts:56]: Verify ISO standard for negative timestamp

File: assets/javascripts/redmine-tracky.js
Comment (CuddlyBunion341) [inline assets/javascripts/redmine-tracky.js:1]: Idea: Reduce JS

File: config/routes.rb [hunk 10]
```diff
@@ -9,8 +9,6 @@
 post 'timer_sessions_rebalance/:id', to: 'timer_sessions#rebalance', as: :timer_sessions_rebalance
 post 'timer_sessions_continue/:id', to: 'timer_sessions#continue', as: :timer_sessions_continue
 
-post 'time_tracker/start', to: 'time_tracker#start', as: :start_time_tracker
-post 'time_tracker/stop', to: 'time_tracker#stop', as: :stop_time_tracker
-post 'time_tracker/update', to: 'time_tracker#update', as: :update_time_tracker
+post 'time_tracker/upsert', to: 'time_tracker#upsert', as: :time_tracker_upsert
```
Comment (CuddlyBunion341) [inline config/routes.rb:12]: Update route

File: app/views/timer_sessions/_timer_container.html.erb [hunk 11]
```diff
@@ -17,80 +17,72 @@
...
-            <%= f.text_field :issue_id, label: false, type: :search, data: { issue_completion_target: 'input' },
-                required: active_timer_session.persisted?, autofocus: 'autofocus' %>
-            <% if SettingsManager.visible_hints %>
-              <small class="form-text text-muted"><%= t('timer_sessions.messaging.issue_list_hint') %></small>
-            <% end %>
-          </div>
-        </div>
+      <div class="mb-3">
+        <%= f.text_field :comments, data: { form_target: 'description', timer_target: 'description', action: 'change->form#change' },
+            placeholder: t('timer_sessions.messaging.comments_placeholder'), class: 'w-100',
```
Comment (CuddlyBunion341) [inline app/views/timer_sessions/_timer_container.html.erb:22]: Maybe hide date at input field

File: assets.src/src/redmine-tracky/controllers/timer-controller.ts (hunk 9 above)
Comment (CuddlyBunion341) [reply assets.src/src/redmine-tracky/controllers/timer-controller.ts:56]: According to the ISO_8601 standard, there seems to be a notation, where the offset is perfixed with a 'minus' sign to imply negative time:
https://en.wikipedia.org/wiki/ISO_8601#Time_offsets_from_UTC

File: app/components/timer_session_entry_component.rb [hunk 12]
```diff
@@ -25,15 +23,15 @@ def timer_session_entry?
   end
 
   def can_destroy_entry?
-    permission_manager.can?(:destroy, :timer_sessions)
+    User.current.allowed_to_globally?(action: :destroy, controller: 'timer_sessions')
```
[the remark this answers is not in the document]
Comment (CuddlyBunion341) [reply app/components/timer_session_entry_component.rb:26]: Maybe in another PR...

**Review by schmijos — CHANGES_REQUESTED** [review]
Please extract the input form re-arrangements to a new PR. I'd like to be able to roll-back the condensed form view if user feedback is bad.

**Comment by CuddlyBunion341** [comment]
Thanks for the review @schmijos! I should have updated the Pull Request description explaining that the changes to the tracky layout have been reverted. You should also be able to see only the paths changed in the views and nothing else. I will open another PR with the requested changes.

File: app/controllers/time_tracker_controller.rb [hunk 13]
```diff
@@ -1,79 +1,92 @@
...
-  def start
-    # Early return if timer already running
-    if @current_timer_session
-      render :update, layout: false and return
+  before_action :set_current_timer_session, only: %i[create_or_update cancel]
+
+  def create_or_update
+    return start_timer unless @current_timer_session
+
+    cancel_timer if params[:cancel].present?
```
Comment (CuddlyBunion341) [inline app/controllers/time_tracker_controller.rb:11]:
```suggestion
```

**Review by schmijos — APPROVED** [review]
Looks good! Let's make sure we test this properly on develop.

File: app/models/unfinished_timer_session_validator.rb [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+class UnfinishedTimerSessionValidator < ActiveModel::Validator
+  def validate(record)
+    @record = record
+
+    validate_start_before_end_date
+  end
+
+  def validate_start_before_end_date
+    return if @record.timer_start.blank? || @record.timer_end.blank?
```
[the remark this answers is not in the document]
Comment (CuddlyBunion341) [reply app/models/unfinished_timer_session_validator.rb:11]: True, thanks for pointing it out

File: app/models/finished_timer_session_validator.rb [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+    @record.errors.add(:timer_start,
+                       :too_short)
+  end
+
+  def validate_day_limit
+    return unless (
+        @record.splittable_hours + TimerSession.created_by(@record.user)
+        .recorded_on(@record.timer_start.to_date).sum(:hours)
+      ) > SettingsManager.max_hours_recorded_per_day.to_f
```
[the remark this answers is not in the document]
Comment (CuddlyBunion341) [reply app/models/finished_timer_session_validator.rb:47]: More readable for sure.

File: app/models/unfinished_timer_session_validator.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (schmijos) [reply app/models/unfinished_timer_session_validator.rb]: Why wasn't this catched by a test?

[the remark this answers is not in the document]
Comment (CuddlyBunion341) [reply app/models/unfinished_timer_session_validator.rb]: The validations have been tested indirectly by the request tests. The timer_start should be set in all scenarios.

**Pull request merged**